File: java/src/demo/AveragePairWithPointer.java
package demo;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class AveragePairWithPointer {

    public static void main(String[] args) {
        int[] arr = {-21, -11, 0, 1, 2, 3, 9, 14, 17, 21, 24};
        double avg = 1.5;
        System.out.println(averagePair(arr,avg));
    }

    public static List<List<Integer>> averagePair(int[] arr, double avg){
        int left = 0;
        int right = arr.length - 1;
        List<List<Integer>> ans = new ArrayList<>();
        while(right > left){
            double val = (double) (arr[left] + arr[right]) / 2;
            if(avg > val){
                left ++;
            }else if(avg < val){
                right --;
            }else {
                ans.add(Arrays.asList(arr[left],arr[right]));
                right --;
                left ++;
            }
        }
        return ans;
    }
}

File: js/algorithm-design/averagePairWithPointer.js
// 找出所有一對一隊的平均值跟給予的值相同
const averagePair = (nums, avg) => {
  let left = 0;
  let right = nums.length - 1;
  let ans = [];
  let count = 0;
  while (right > left) {
    let val = (nums[left] + nums[right]) / 2;
    if (val > avg) {
      right--;
      count++;
    } else if (val < avg) {
      left++;
    } else if (val === avg) {
      ans.push([nums[left], nums[right]]);
      left++;
      right--;
      count++;
    }
  }
  console.log("count-", count);
  return ans;
};

const arr = [-21, -11, 0, 1, 2, 3, 9, 14, 17, 21, 24];
console.log(averagePair(arr, 1.5));
